### PR TITLE
fix #4574, #4575 feat(visualization): group default secondary metrics 

### LINKED ISSUE #4574: Group search metrics & usage metrics server-side
Right now, there is a semi-hard-coded mapping from a few different metrics to their statistic that should be used to graph it in the front-end:
https://github.com/mozilla/experimenter/blob/main/app/experimenter/visualization/api/v3/views.py#L50

In a similar fashion, we want the server to specify groups of metrics that should be "collapsed" together in the front-end. Specifically, we want to collapse together all search metrics (E.g. "tagged follow on search count, tagged search count, etc) and also to collapse usage metrics such as active hours/ad clicks/uri counts.

Here is a sample experiment with these metrics listed: https://experimenter.services.mozilla.com/nimbus/onoff-aboutwelcome-experiment/results

On the server side, the implementation just includes creating this mapping/grouping for the 2 groups and including it in the API response.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-929)
┆Issue Number: EXP-929

### PR DESCRIPTION
**Pull request by robhudson**
Screenshot 2021-05-25 at 16-32-37 On Off about welcome Experiment – Analysis Experimenter

### CODE REVIEW

File: app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx [hunk 1]
```diff
@@ -190,7 +191,8 @@ const formattedAnalysisPointComparator = (
 
 const TableVisualizationRow: React.FC<{
   metricKey: string;
-  results: BranchDescription;
+  results: BranchDescription | BranchDescriptionGrouped;
```
Comment (emtwo) [inline app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx:194]: When are the results not grouped?

File: app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx [hunk 2]
```diff
@@ -221,13 +224,15 @@ const TableVisualizationRow: React.FC<{
     className = "";
     tooltipText = tooltip;
     field = <div>{BASELINE_TEXT}</div>;
-    const percent = branch_data[METRIC.USER_COUNT]["percent"];
+    const percent = branch_data["other"][METRIC.USER_COUNT]["percent"];
```
Comment (emtwo) [inline app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx:227]: `other` should be a constant

File: app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx [hunk 3]
```diff
@@ -221,13 +224,15 @@ const TableVisualizationRow: React.FC<{
...
     const branchType = is_control ? VARIANT_TYPE.CONTROL : VARIANT_TYPE.VARIANT;
     branchComparison =
       branchComparison || dataTypeMapping[tableLabel][branchType];
 
     const userCountsList =
-      branch_data[METRIC.USER_COUNT][BRANCH_COMPARISON.ABSOLUTE]["all"];
+      branch_data["other"][METRIC.USER_COUNT][BRANCH_COMPARISON.ABSOLUTE][
```
Comment (emtwo) [inline app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx:233]: `other` should be a constant

File: app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx [hunk 4]
```diff
@@ -221,13 +224,15 @@ const TableVisualizationRow: React.FC<{
...
+        "all"
```
Comment (emtwo) [inline app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx:234]: arguably `all` should be a constant too but it's not part of this PR

File: app/experimenter/nimbus-ui/src/lib/visualization/utils.tsx [hunk 5]
```diff
@@ -49,21 +48,24 @@ export const getTableDisplayType = (
 };
 
 export const getSortedBranches = (analysis: AnalysisData) => {
-  const results: { [branch: string]: BranchDescription } | null =
-    analysis?.overall || analysis?.weekly;
-  if (!results) {
-    return [];
+  if (analysis?.overall) {
+    return Object.keys(analysis.overall)
+      .filter((b) => analysis.overall?.[b].is_control)
```
Comment (emtwo) [inline app/experimenter/nimbus-ui/src/lib/visualization/utils.tsx:53]: nit: can we use the full variable name `branch` instead of `b`?

File: app/experimenter/nimbus-ui/src/components/PageResults/index.tsx [hunk 6]
```diff
@@ -119,21 +127,65 @@ const PageResults: React.FunctionComponent<RouteComponentProps> = () => {
                       results={analysis}
                       outcomeSlug={outcome!.slug!}
                       outcomeDefaultName={outcome!.friendlyName!}
+                      group="other"
```
Comment (emtwo) [inline app/experimenter/nimbus-ui/src/components/PageResults/index.tsx:130]: `other` as a constant

**Review by emtwo — COMMENTED** [review]
Code looks great overall! Thanks for your patience with this!

I left some comments but they're all minor. I'm going to test it out with a few different experiments before approving but likely next week.

File: app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx (hunk 1 above)
Comment (robhudson) [reply app/experimenter/nimbus-ui/src/components/PageResults/TableVisualizationRow/index.tsx:194]: Yep, I will fix that. Once I moved to making the weekly graphs accept a group argument this should no longer exist.

File: app/experimenter/visualization/api/v3/models.py [hunk 7]
```diff
@@ -33,6 +33,34 @@ class Segment:
     ALL = "all"
 
 
+SEARCH_METRICS = [
+    "searches_with_ads",
+    "search_count",  # Remove this when above is done.
```
[the remark this answers is not in the document]
Comment (robhudson) [reply app/experimenter/visualization/api/v3/models.py:38]: I think the code moved. It's now below. :) I was referring to the guardrail grouping.

**Comment by emtwo** [comment]
In visually reviewing the PR, I noticed a couple of things:

1) I think it should say "Show Other Metrics" instead of "Show other". Just a little clearer (screenshot below)
Screen Shot 2021-06-07 at 12 54 26 PM

2) This one is a little bigger of an issue. In the screenshot below, the left navbar shows the default metrics but they are collapsed by default. This is confusing but also the bigger issue is that clicking on one of the links in the nav on the left does not scroll you down to where the metric is. Perhaps the metrics should not show up in the nav on the left if they are collapsed. It would also be nice to have subheadings in the left nav to show what the groups are instead of them all being under "default metrics"
Screen Shot 2021-06-07 at 12 54 43 PM

I realize the changes for #2 might be a lot. I think what we could do is make the collapse open by default for now so we can merge this PR. Then in a follow-up PR, the sidebar changes can be made and then the collapse would be changed to closed by default. @robhudson let me know what you think.

File: app/experimenter/visualization/api/v3/models.py [hunk 8]
```diff
@@ -235,16 +272,36 @@ def create_results_object_model(data):
...
 
     class Branch(BaseModel):
         is_control: bool = False
         branch_data: BranchData
 
     for branch in branches:
-        branches[branch] = Branch(is_control=False, branch_data=BranchData())
+        branches[branch] = Branch(
+            is_control=False,
+            branch_data=BranchData(
```
[the remark this answers is not in the document]
Comment (robhudson) [reply app/experimenter/visualization/api/v3/models.py:299]: If I take it out I get a test error since I added a new `getattr` call in the view that is looking for "other_metrics".

**Review by emtwo — APPROVED** [review]
this is amazing!!! thanks so much and am excited to see it live 
image
